### PR TITLE
v0.12 Upgrades Back End

### PR DESCRIPTION
**Pull request by aumu1031**

### CODE REVIEW

File: classes/village/VillageManager.php
Comment (aumu1031) [inline classes/village/VillageManager.php:1729]: Will shift this entire code block and sub-functions to `ProposalManager` and use `ActionResult` instead of try/catch

File: classes/village/VillageManager.php
Comment (aumu1031) [inline classes/village/VillageManager.php:2511]: Only trade towns in a region if they're currently controlled.

File: classes/Village.php
Comment (aumu1031) [inline classes/Village.php:1]: Thinking we can turn `activeUpgradeEffects[]` into a sub-class like `$village->policy` instead of having to get upgrade effects by key. So that way all of the upgrade/building configuration will be accessed via classes rather than needing to interact directly with anything in the configuration file.

File: classes/Village.php
Comment (aumu1031) [inline classes/Village.php:67]: This policy benefits to research/construction are based on how many villages have more total research/construction.

File: classes/Village.php
Comment (aumu1031) [inline classes/Village.php:98]: Need to combine this constructor logic.

File: classes/village/VillageManager.php
Comment (aumu1031) [inline classes/village/VillageManager.php:550]: This changes to a single query rather than multiple queries for multiple resources. Used only for the village HQ display.

File: classes/village/VillageUpgradeManager.php
Comment (aumu1031) [inline classes/village/VillageUpgradeManager.php:719]: These are a count of total days completed construction/research time, currently just for the catch-up bonus with Ashes policy but also a useful way for us to glance at the DB and see how much progress villages are making.

File: classes/Village.php
Comment (levimeahan) [reply classes/Village.php:1]: sounds good!

File: classes/Village.php
Comment (levimeahan) [inline classes/Village.php:77]: Consider extracting this to a method and then doing something like `$this->construction_speed = $this->getConstructionSpeed()`

File: classes/Village.php
Comment (levimeahan) [inline classes/Village.php:104]: one other note, we're making two separate queries to `villages` here, we could consider combining this into one query with the field that `construction_speed` and `research_speed` calculations use and pre-loading it into this class

File: classes/village/VillageUpgradeManager.php
Comment (levimeahan) [inline classes/village/VillageUpgradeManager.php:162]: Hmm, I wonder if there's a case for us to use the raw time value. Could consider storing the raw value in the DTO and then format it to a string where needed (can be a helper method on the `VillageUpgradeDTO` / `VillageUpgrade` class)

**Review by levimeahan — APPROVED** [review]
`ui_components/build` files are not supposed to be in the PR right? I'm okay merging as is and waiting on deploy to dev until the other PRs are merged

**Pull request merged**
